webui/react/src/hooks/useSettings.ts

```diff
@@ -1,359 +1,299 @@
 };
 
-export const queryParamToType = (type: BaseType, param: string | null): Primitive | undefined => {
-  if (param == null) return undefined;
+const queryParamToType = (type: BaseType, param: string | null): Primitive | undefined => {
+  if (param === null || param === undefined) return undefined;
   if (type === BaseType.Boolean) return param === 'true';
   if (type === BaseType.Float || type === BaseType.Integer) {
-    const value = type === BaseType.Float ? parseFloat(param) : parseInt(param);
+    const value = Number(param);
     return !isNaN(value) ? value : undefined;
   }
   if (type === BaseType.Object) return JSON.parse(param);
-
   if (type === BaseType.String) return param;
   return undefined;
 };
 
-export const queryToSettings = <T>(config: SettingsConfig, query: string): T => {
+const queryToSettings = <T>(config: SettingsConfig<T>, query: string) => {
   const params = queryString.parse(query);
-  return config.settings.reduce((acc, prop) => {
-    /*
-     * Attempt to decode the query parameter and if anything
-     * goes wrong, set it to the default value.
-     */
-    try {
-      const paramValue = params[prop.key];
-      const baseType = prop.type.baseType;
 
+  return (Object.values(config.settings) as SettingsConfigProp<typeof config>[]).reduce<Settings>(
+    (acc, setting) => {
       /*
-       * Convert the string-based query params to primitives.
-       * `undefined` values can happen if the query param values are invalid.
-       *   string[] => Primitive[] | undefined
-       *   string   => Primitive | undefined
-       *   null     => undefined
+       * Attempt to decode the query parameter and if anything
+       * goes wrong, set it to the default value.
        */
-      const queryValue = Array.isArray(paramValue)
-        ? paramValue
-            .map((value) => queryParamToType(baseType, value))
-            .filter((value): value is Primitive => value !== undefined)
-        : queryParamToType(baseType, paramValue);
+      try {
+        const paramValue = params[setting.storageKey];
+        const baseType = setting.type;
+        const expectingArray = baseType.is([]);
+
+        if (paramValue !== null) {
+          let queryValue: Primitive | Primitive[] | undefined;
+          /*
+           * Convert the string-based query params to primitives.
+           * `undefined` values can happen if the query param values are invalid.
+           *   string[] => Primitive[] | undefined
+           *   string   => Primitive | undefined
+           */
+          if (Array.isArray(paramValue)) {
+            queryValue = paramValue.reduce<Primitive[]>((acc, value) => {
+              const parsedValue = queryParamToType(baseType.name as BaseType, value);
+
+              if (parsedValue !== undefined) acc.push(parsedValue);
+
+              return acc;
+            }, []);
+          } else {
+            queryValue = queryParamToType(baseType.name as BaseType, paramValue);
```

Comment:
You should be using the `decode` in io-ts to actually parse out the query params. You may need a special case for strings if you want to keep quotes out of them.
Something like:
```typescript
if (baseType === t.string) { // optional string handling is left as an exercise to the reader
   queryValue = baseType.decode(paramValue)
} else {
   queryValue = baseType.decode(JSON.parse(paramValue))
}
```

Reply:
`foo.decode` does indeed do what we want here. `t.number.decode(5)` will return `Right(5)`, `t.number.decode('str')` will return `Left(...)`.